README.md
Job Configuration History Plugin
==============================
[![Build Status](https://ci.jenkins.io/buildStatus/icon?job=Plugins/jobConfigHistory-plugin/master)](https://ci.jenkins.io/job/plugins/job/jobConfigHistory-plugin/)
[![Jenkins Plugin](https://img.shields.io/jenkins/plugin/v/jobConfigHistory.svg)](https://plugins.jenkins.io/jobConfigHistory)
[![Jenkins Plugin Installs](https://img.shields.io/jenkins/plugin/i/jobConfigHistory.svg?color=blue)](https://plugins.jenkins.io/jobConfigHistory)
[![Contributors](https://img.shields.io/github/contributors/jenkinsci/jobConfigHistory-plugin.svg)](https://github.com/jenkinsci/job-config-history-plugin/graphs/contributors)

Saves copies of all job and system configurations.

Older versions of this plugin may not be safe to use. Please review the following warnings before using an older version:
* [Reflected Cross Site Scripting vulnerability](https://jenkins.io/security/advisory/2018-09-25/#SECURITY-1130)

This plugin saves **a copy of the configuration file** of jobs and agents (`config.xml`) for every change made and of the system configuration (`<config-name>.xml`).
You can also see what changes have been made by which user if you configured a security policy.

It is also possible to get a **side-by-side view** of the differences between two configurations and to restore an old version of a job's configuration. (The latter is only available for jobs, not for system changes.)
However, if you restore an older version of the config file and the new version contains fields that were not present in the older version, the restored version will still contain these fields, although they were not present in the original.

The plugin also provides an **overview page** of all changes. You can find it under `<jenkins_url>/jobConfigHistory` or reach it via links in the sidepanel of the main and the system configuration pages.
The overview page initially only lists system configuration changes (for performance reasons), but there are links to view either all job configuration histories or just the deleted jobs or all kinds of configuration history entries together (which may take some time to load, depending on the number of jobs in your instance).

------------------------------------------------------------------------

# Pictures

#### Job Config History Revision Overview

![](docs/img/Job_Config_History_Revision_Overview.png)

#### Job Diff Side-By-Side View

![](docs/img/Job_Diff_Side-By-Side_View.png) 

#### Config Overview Page

![](docs/img/Config_Overview_Page.png)


------------------------------------------------------------------------

# Interdependencies with other Plugins/ Excluding config path parts (jobs, folders, ...)

If you use other plugins, that are (automatically) changing the configuration, you might end up with a lot of unwanted change detections.

It's recommended to use the given ability, to exclude such changes by patterns. This can be done in the System Configuration or via `<jenkins-link>/configure`. Just add the corresponding pattern (e.g. cluster-stats) in the input **Configuration exclude file pattern**:

![](docs/img/globalconfig.png)

This functionality also allows you to exclude jobs, folders or other parts of your job hierarchy:
The pattern is attempted to be found in a config file's file path. Adding `|test` to the pattern will also cause `testFolder/jobs/job1` to be excluded.
So if you want to exclude jobs, only, it might be better to enter something like `|testJob/config\.xml`.

If you use this plugin together with the [Global Build Stats Plugin](https://plugins.jenkins.io/Global+Build+Stats+Plugin) you should add `global-build-stats` to the excludes in the **Configuration exclude file pattern** section of this plugin in **Configure System** page. Fixed in version 1.9 of the plugin.

If you use this plugin together with the [Cluster Statistics Plugin](https://plugins.jenkins.io/Cluster+Statistics+Plugin) you should add `cluster-stats` to the excludes in the **Configuration exclude file pattern** section of this plugin in **Configure System** page. Otherwise it might exceed your storage.

------------------------------------------------------------------------

# Open Issues

* See [open issues](https://issues.jenkins.io/browse/JENKINS-66956?jql=component%20%3D%20jobconfighistory-plugin) on issues.jenkins-ci.org.

------------------------------------------------------------------------

# Changelog

See
[Changelog](https://github.com/jenkinsci/job-config-history-plugin/releases)
on Github.

------------------------------------------------------------------------
# Notes

* This plugin uses *highlight.js* for syntax highlighting (code and diffs). See
    + [github](https://github.com/highlightjs/highlight.js/)
    + [homepage](https://highlightjs.org/)
    
* Development: Make `hpi:run` work with shared libraries
    * install shared library plugin (if not installed already)
    * install git (if you want to use the shared library plugin with git)


CHANGELOG.md
# Changelog moved to [GitHub Releases](https://github.com/jenkinsci/job-config-history-plugin/releases)

## 2.28.1 (July 16 2021)
* revert [PR #129](https://github.com/jenkinsci/job-config-history-plugin/pull/129)

## 2.28 (July 2 2021)
* [PR #126](https://github.com/jenkinsci/job-config-history-plugin/pull/126) fix biggest style errors
* [PR #129](https://github.com/jenkinsci/job-config-history-plugin/pull/129) remove Jenkins.get().getRootUrl() to generate relative url

## 2.27 (April 7 2021)
* [PR #127](https://github.com/jenkinsci/job-config-history-plugin/pull/127) [JEP-227] Prepare plugin for Acegi update

## 2.26 (April 30 2020)
* [JENKINS-61049](https://issues.jenkins-ci.org/browse/JENKINS-61049) Move documentation from jenkins wiki to github
* [PR #118](https://github.com/jenkinsci/job-config-history-plugin/pull/118) Simplify code
* [PR #123](https://github.com/jenkinsci/job-config-history-plugin/pull/123) Remove debug lines

## 2.25 (March 13 2020)
* Remove unused HTML help files (PR #117) Thanks, darxriggs 
* Reduce deprecated API usage (PR #116) Thanks, darxriggs 
* Upgrade parent pom and dependencies (PR #115) Thanks, darxriggs 
* Fix Javadoc errors (PR #114) Thanks, darxriggs 
* (JENKINS-6659) configChange-comment
* (JENKINS-60163) Introducing paging
* (JENKINS-58601) privileges management
* (JENKINS-59812) config history provide helpful error when it can't create history
* (JENKINS-58141) Startup issues
* Fix indentation with multi-line lines.

## Version 2.24 (September 9 2019)
* (PR #96) Modernize overall design (mostly adjust it to Pipeline Configuration History).
* (PR #100) [JENKINS-58601] Allow deletion of config revisions in the respective config history overview.
* (PR #101) [JENKINS-58316] Improve error message for XmlFile::write RuntimeExceptions.
* (PR #102)
* Fix empty nodes not being displayed in diff view.
* Add xml indentation.
* Update DiffUtils.
* (PR #105) [JENKINS-55787] Switch labels from entry to checkbox (Thanks,  Unknown User (jsoref))

## Version 2.23.1 (August 6 2019)
* (PR #97) Bugfix: System configs were not persisted (regression of 8e2d9dd)

## Version 2.23 (July 23 2019)
* (PR #90) Allow user to use ignore job config history by regex. (Thanks, Unknown User (fengxx))
* (PR #91) Trim numeric form values (incl. validation) (Thanks, René Scheibe)

## Version 2.22 (June 3 2019)
* (PR #88) Code Improvements. (Thanks, René Scheibe)
* (PR #89) Init Level below COMPLETED might use SYSTEM (Thanks, Emilio Escobar )
* Version 2.21 (May 2 2019)
* (PR #87) Modernize (parent pom and Jenkins). (Thanks, René Scheibe)

## Version 2.20 (April 5 2019)
* (JENKINS-41177) Identical job configuration can have different config.xml based on edit method.
* (PR #81) Hide version changes
* (JENKINS-56023, PR #83) Display folder jobs in root history overview correctly.
* (JENKINS-45821, PR #74) Add the MilestoneStep file by default
* (JENKINS-49301, PR #84) Move job config history when a job is moved.

## Version 2.19 (December 5 2018)
* (JENKINS-53998) Job Config History plugin should not call User.current() during startup
* (PR #72) Use canonical user identify in case change initiator is unknown
* (PR #67) Fix root URL
* Findbugs fixes 
* Minor changes (Jenkinsfile, Link fixes)

## Version 2.18.3 (October 25 2018)
* (JENKINS-54223) Job configurations displayed incorrectly (raw)

## Version 2.18.2 (September 28 2018)
* (JENKINS-53819) Job configurations displayed incorrectly

## Version 2.18.1 (September 25 2018)
* Fix security issue

## Version 2.18 (September 25 2017)
* (JENKINS-47089) Wiki page for JobConfigHistory has circular link
* (PR #65) Also exclude files from the lockable resources plugin

## Version 2.17 (August 22 2017)
* (JENKINS-21600) Fixed max number of history entries issue.
* (JENKINS-19141) Make maven-plugin optional.

## Version 2.16 (April 21 2017)
* (JENKINS-42464) Job Configuration History not accessible from public IP address
* Switched unit tests from HudsonTestCase to JenkinsRule.

## Version 2.15 (July 21 2016)
* (JENKINS-34802) Exclude history collection for selected users (Note: All changes will be assigned to the next, not excluded user!)

## Version 2.14 (May 09 2016)
* (JENKINS-33641) Showing old and new job name if hovering over info image beneath operation renamed (for jobs and nodes)
* (JENKINS-25654) Fixed: Restore Project option for Job Config History gets Oops! page
* Merged pull request #49 (thanks to Brandon Koepke)
* (JENKINS-34151) Add support for Pipeline projects to display config changes in build history
* (JENKINS-30578) Error when trying to view job config history

## Version 2.13 (March 18 2016)
* (JENKINS-33289) NPE when clicking showDiffs (Jenkins Core 1.650 or higher)
* SECURITY-140 XSS vulnerability

## Version 2.12 (July 23 2015)
* (JENKINS-29063) Switch default for per-Maven modules to false (Thanks to Andrew Bayer)
* (JENKINS-24930) Alternating row bg colors (Thanks to Daniel Beck)

## Version 2.11 (April 17 2015)
* Avoid calling User.current() during Jenkins initialization (Thanks to Thomas de Grenier de Latour)
* Avoid tracking changes for cloud agents as well (Thanks to Ryan Campbel)

## Version 2.10 (November 12 2014)
* Fix proposal for NPE in ComputerHistoryListener.onConfigurationChange method (Thanks to William Bernardet)
* (JENKINS-22639) Don't record changes to AbstractCloudSlaves or Ephemeral Nodes (Thanks to Ryan Campbell and Jesse Click)

## Version 2.9 (September 02 2014)
* Fixed: Do not save config for matrix configurations (JENKINS-24412) (Thanks to Oliver Gondža)
* Fixed: Job config change not saved for maven and matrix projects (JENKINS-24410) (Thanks to Oliver Gondža)

## Version 2.8 (July 17 2014)
* Default for global configuration "saveModuleConfiguration" changed to false
* Fixed history could not saved for some job types if in the system configuration "Save folder configuration changes" wasn't set (Thanks to Jesse Glick)

src/main/java/bmsi/util/DiffPrint.java
/*
 * Original: http://www.bmsi.com/java/DiffPrint.java
 *
 * $Log: DiffPrint.java,v $
 * Revision 1.10  2009/04/20 18:53:29  stuart
 * Make print_header public.
 *
 * Revision 1.9  2009/01/19 03:05:26  stuart
 * Fix StackOverflow bug with heuristic on reported by Jimmy Han.
 *
 * Revision 1.8  2007/12/20 05:04:21  stuart
 * Can no longer import from default package.
 *
 * Revision 1.7  2007/12/20 04:29:53  stuart
 * Fix setupOutput permission.  Thanks to Sargon Benjamin
 *
 * Revision 1.6  2005/04/27 02:13:40  stuart
 * Add Str.dup()
 *
 * Revision 1.5  2004/01/29 02:35:35  stuart
 * Test for out of bounds exception in UnifiedPrint.print_hunk.
 * Add setOutput() to DiffPrint.Base.
 *
 * Revision 1.4  2003/04/22  01:50:47  stuart
 * add Unified format diff
 *
 * Revision 1.3  2003/04/22  01:00:32  stuart
 * added context diff format
 *
 * Revision 1.2  2000/03/02  16:59:54  stuart
 * add GPL
 *
 */
package bmsi.util;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.io.Writer;
import java.util.Date;
import java.util.Vector;

interface UnaryPredicate {
  boolean execute(Object obj);
}

/** A simple framework for printing change lists produced by <code>Diff</code>.
    @see bmsi.util.Diff
    @author Stuart D. Gathman
    Copyright (C) 2000 Business Management Systems, Inc.
<p>
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 1, or (at your option)
    any later version.
<p>
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
<p>
    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
public class DiffPrint {
  /** A Base class for printing edit scripts produced by Diff.
      This class divides the change list into "hunks", and calls
      <code>print_hunk</code> for each hunk.  Various utility methods
      are provided as well.
   */
  public static abstract class Base {
    protected PrintWriter outfile;
    public void setOutput(Writer wtr) {
      outfile = new PrintWriter(wtr);
    }
    protected void setupOutput() {
      if (outfile == null)
	outfile = new PrintWriter(new OutputStreamWriter(System.out));
    }
    protected Base(Object[] a,Object[] b) {
      file0 = a;
      file1 = b;
    }
    /** Set to ignore certain kinds of lines when printing
	an edit script.  For example, ignoring blank lines or comments.
     */
    protected UnaryPredicate ignore = null;

    /** Set to the lines of the files being compared.
     */
    protected Object[] file0, file1;

    /** Divide SCRIPT into pieces by calling HUNKFUN and
       print each piece with PRINTFUN.
       Both functions take one arg, an edit script.

       PRINTFUN takes a subscript which belongs together (with a null
       link at the end) and prints it.  */
    public void print_script(Diff.change script) {
      setupOutput();
      Diff.change next = script;

      while (next != null)
	{
	  Diff.change t, end;

	  /* Find a set of changes that belong together.  */
	  t = next;
	  end = hunkfun(next);

	  /* Disconnect them from the rest of the changes,
	     making them a hunk, and remember the rest for next iteration.  */
	  next = end.link;
	  end.link = null;
	  //if (DEBUG)
	  //  debug_script(t);

	  /* Print this hunk.  */
	  print_hunk(t);

	  /* Reconnect the script so it will all be freed properly.  */
	  end.link = next;
	}
	outfile.flush();
    }

    /** Called with the tail of the script
       and returns the last link that belongs together with the start
       of the tail. */

    protected Diff.change hunkfun(Diff.change hunk) {
      return hunk;
    }

    protected int first0, last0, first1, last1, deletes, inserts;

    /** Look at a hunk of edit script and report the range of lines in each file
      that it applies to.  HUNK is the start of the hunk, which is a chain
      of `struct change'.  The first and last line numbers of file 0 are stored
      in *FIRST0 and *LAST0, and likewise for file 1 in *FIRST1 and *LAST1.
      Note that these are internal line numbers that count from 0.

      If no lines from file 0 are deleted, then FIRST0 is LAST0+1.

      Also set *DELETES nonzero if any lines of file 0 are deleted
      and set *INSERTS nonzero if any lines of file 1 are inserted.
      If only ignorable lines are inserted or deleted, both are
      set to 0.  */

    protected void analyze_hunk(Diff.change hunk) {
      int f0, l0 = 0, f1, l1 = 0, show_from = 0, show_to = 0;
      int i;
      Diff.change next;
      boolean nontrivial = (ignore == null);

      show_from = show_to = 0;

      f0 = hunk.line0;
      f1 = hunk.line1;

      for (next = hunk; next != null; next = next.link)
	{
	  l0 = next.line0 + next.deleted - 1;
	  l1 = next.line1 + next.inserted - 1;
	  show_from += next.deleted;
	  show_to += next.inserted;
	  for (i = next.line0; i <= l0 && ! nontrivial; i++)
	    if (!ignore.execute(file0[i]))
	      nontrivial = true;
	  for (i = next.line1; i <= l1 && ! nontrivial; i++)
	    if (!ignore.execute(file1[i]))
	      nontrivial = true;
	}

      first0 = f0;
      last0 = l0;
      first1 = f1;
      last1 = l1;

      /* If all inserted or deleted lines are ignorable,
	 tell the caller to ignore this hunk.  */

      if (!nontrivial)
	show_from = show_to = 0;

      deletes = show_from;
      inserts = show_to;
    }

    /** Called to print the script header which identifies the files compared.
      The default does nothing (except set output to system.out if
      not otherwise set).  Derived style classes can override to print
      the files compared in the format for that style.
     */
    public void print_header(String filea, String fileb) {
      setupOutput();
    }

    protected abstract void print_hunk(Diff.change hunk);

    protected void print_1_line(String pre,Object linbuf) {
      outfile.println(pre + linbuf.toString());
    }

    /** Print a pair of line numbers with SEPCHAR, translated for file FILE.
       If the two numbers are identical, print just one number.

       Args A and B are internal line numbers.
       We print the translated (real) line numbers.  */

    protected void print_number_range (char sepchar, int a, int b) {
      /* Note: we can have B < A in the case of a range of no lines.
	 In this case, we should print the line number before the range,
	 which is B.  */
      if (++b > ++a)
	outfile.print("" + a + sepchar + b);
      else
	outfile.print(b);
    }

    public static char change_letter(int inserts, int deletes) {
      if (inserts == 0)
	return 'd';
      else if (deletes == 0)
	return 'a';
      else
	return 'c';
    }
  }

  /** Print a change list in the standard diff format.
   */
  public static class NormalPrint extends Base {

    public NormalPrint(Object[] a,Object[] b) {
      super(a,b);
    }

    /** Print a hunk of a normal diff.
       This is a contiguous portion of a complete edit script,
       describing changes in consecutive lines.  */

    protected void print_hunk (Diff.change hunk) {

      /* Determine range of line numbers involved in each file.  */
      analyze_hunk(hunk);
      if (deletes == 0 && inserts == 0)
	return;

      /* Print out the line number header for this hunk */
      print_number_range (',', first0, last0);
      outfile.print(change_letter(inserts, deletes));
      print_number_range (',', first1, last1);
      outfile.println();

      /* Print the lines that the first file has.  */
      if (deletes != 0)
	for (int i = first0; i <= last0; i++)
	  print_1_line ("< ", file0[i]);

      if (inserts != 0 && deletes != 0)
	outfile.println("---");

      /* Print the lines that the second file has.  */
      if (inserts != 0)
	for (int i = first1; i <= last1; i++)
	  print_1_line ("> ", file1[i]);
    }
  }

  /** Prints an edit script in a format suitable for input to <code>ed</code>.
      The edit script must be generated with the reverse option to
      be useful as actual <code>ed</code> input.
   */
  public static class EdPrint extends Base {

    public EdPrint(Object[] a,Object[] b) {
      super(a,b);
    }

    /** Print a hunk of an ed diff */
    protected void print_hunk(Diff.change hunk) {

      /* Determine range of line numbers involved in each file.  */
      analyze_hunk (hunk);
      if (deletes == 0 && inserts == 0)
	return;

      /* Print out the line number header for this hunk */
      print_number_range (',', first0, last0);
      outfile.println(change_letter(inserts, deletes));

      /* Print new/changed lines from second file, if needed */
      if (inserts != 0)
	{
	  boolean inserting = true;
	  for (int i = first1; i <= last1; i++)
	    {
	      /* Resume the insert, if we stopped.  */
	      if (! inserting)
		outfile.println(i - first1 + first0 + "a");
	      inserting = true;

	      /* If the file's line is just a dot, it would confuse `ed'.
		 So output it with a double dot, and set the flag LEADING_DOT
		 so that we will output another ed-command later
		 to change the double dot into a single dot.  */

	      if (".".equals(file1[i]))
		{
		  outfile.println("..");
		  outfile.println(".");
		  /* Now change that double dot to the desired single dot.  */
		  outfile.println(i - first1 + first0 + 1 + "s/^\\.\\././");
		  inserting = false;
		}
	      else
		/* Line is not `.', so output it unmodified.  */
		print_1_line ("", file1[i]);
	    }

	  /* End insert mode, if we are still in it.  */
	  if (inserting)
	    outfile.println(".");
	}
    }
  }

  /** Prints an edit script in context diff format.  This and its
    'unified' variation is used for source code patches.
   */
  public static class ContextPrint extends Base {

    protected int context = 3;

    public ContextPrint(Object[] a,Object[] b) {
      super(a,b);
    }

    protected void print_context_label (String mark, File inf, String label) {
      setupOutput();
      if (label != null)
	outfile.println(mark + ' ' + label);
      else if (inf.lastModified() > 0)
        // FIXME: use DateFormat to get precise format needed.
	outfile.println(
	  mark + ' ' + inf.getPath() + '\t' + new Date(inf.lastModified())
	);
      else
	/* Don't pretend that standard input is ancient.  */
	outfile.println(mark + ' ' + inf.getPath());
    }

    public void print_header(String filea,String fileb) {
      print_context_label ("***", new File(filea), filea);
      print_context_label ("---", new File(fileb), fileb);
    }

    /** If function_regexp defined, search for start of function. */
    private String find_function(Object[] lines, int start) {
      return null;
    }

    protected void print_function(Object[] file,int start) {
      String function = find_function (file0, first0);
      if (function != null) {
	outfile.print(" ");
	outfile.print(
	  (function.length() < 40) ? function : function.substring(0,40)
	);
      }
    }

    protected void print_hunk(Diff.change hunk) {

      /* Determine range of line numbers involved in each file.  */

      analyze_hunk (hunk);

      if (deletes == 0 && inserts == 0)
	return;

      /* Include a context's width before and after.  */

      first0 = Math.max(first0 - context, 0);
      first1 = Math.max(first1 - context, 0);
      last0 = Math.min(last0 + context, file0.length - 1);
      last1 = Math.min(last1 + context, file1.length - 1);


      outfile.print("***************");

      /* If we looked for and found a function this is part of,
	 include its name in the header of the diff section.  */
      print_function (file0, first0);

      outfile.println();
      outfile.print("*** ");
      print_number_range (',', first0, last0);
      outfile.println(" ****");

      if (deletes != 0) {
	Diff.change next = hunk;

	for (int i = first0; i <= last0; i++) {
	  /* Skip past changes that apply (in file 0)
	     only to lines before line I.  */

	  while (next != null && next.line0 + next.deleted <= i)
	    next = next.link;

	  /* Compute the marking for line I.  */

	  String prefix = " ";
	  if (next != null && next.line0 <= i)
	    /* The change NEXT covers this line.
	       If lines were inserted here in file 1, this is "changed".
	       Otherwise it is "deleted".  */
	    prefix = (next.inserted > 0) ? "!" : "-";

	  print_1_line (prefix, file0[i]);
	}
      }

      outfile.print("--- ");
      print_number_range (',', first1, last1);
      outfile.println(" ----");

      if (inserts != 0) {
	Diff.change next = hunk;

	for (int i = first1; i <= last1; i++) {
	  /* Skip past changes that apply (in file 1)
	     only to lines before line I.  */

	  while (next != null && next.line1 + next.inserted <= i)
	    next = next.link;

	  /* Compute the marking for line I.  */

	  String prefix = " ";
	  if (next != null && next.line1 <= i)
	    /* The change NEXT covers this line.
	       If lines were deleted here in file 0, this is "changed".
	       Otherwise it is "inserted".  */
	    prefix = (next.deleted > 0) ? "!" : "+";

	  print_1_line (prefix, file1[i]);
	}
      }
    }
  }

  /** Prints an edit script in context diff format.  This and its
    'unified' variation is used for source code patches.
   */
  public static class UnifiedPrint extends ContextPrint {

    public UnifiedPrint(Object[] a,Object[] b) {
      super(a,b);
    }

    public void print_header(String filea,String fileb) {
      print_context_label ("---", new File(filea), filea);
      print_context_label ("+++", new File(fileb), fileb);
    }

    private void print_number_range (int a, int b) {
      //translate_range (file, a, b, &trans_a, &trans_b);

      /* Note: we can have B < A in the case of a range of no lines.
	 In this case, we should print the line number before the range,
	 which is B.  */
      if (b < a)
	outfile.print(b + ",0");
      else
        super.print_number_range(',',a,b);
    }

    protected void print_hunk(Diff.change hunk) {
      /* Determine range of line numbers involved in each file.  */
      analyze_hunk (hunk);

      if (deletes == 0 && inserts == 0)
	return;

      /* Include a context's width before and after.  */

      first0 = Math.max(first0 - context, 0);
      first1 = Math.max(first1 - context, 0);
      last0 = Math.min(last0 + context, file0.length - 1);
      last1 = Math.min(last1 + context, file1.length - 1);

      outfile.print("@@ -");
      print_number_range (first0, last0);
      outfile.print(" +");
      print_number_range (first1, last1);
      outfile.print(" @@");

      /* If we looked for and found a function this is part of,
	 include its name in the header of the diff section.  */
      print_function(file0,first0);

      outfile.println();

      Diff.change next = hunk;
      int i = first0;
      int j = first1;

      while (i <= last0 || j <= last1) {

	/* If the line isn't a difference, output the context from file 0. */

	if (next == null || i < next.line0) {
	  if (i < file0.length) {
	    outfile.print(' ');
	    print_1_line ("", file0[i++]);
	  }
	  j++;
	}
	else {
	  /* For each difference, first output the deleted part. */

	  int k = next.deleted;
	  while (k-- > 0) {
	    outfile.print('-');
	    print_1_line ("", file0[i++]);
	  }

	  /* Then output the inserted part. */

	  k = next.inserted;
	  while (k-- > 0) {
	    outfile.print('+');
	    print_1_line ("", file1[j++]);
	  }

	  /* We're done with this hunk, so on to the next! */

	  next = next.link;
	}
      }
    }
  }


  /** Read a text file into an array of String.  This provides basic diff
     functionality.  A more advanced diff utility will use specialized
     objects to represent the text lines, with options to, for example,
     convert sequences of whitespace to a single space for comparison
     purposes.
   */
  @SuppressWarnings("unchecked")
static String[] slurp(String file) throws IOException {
    BufferedReader rdr = new BufferedReader(new FileReader(file));
    Vector s = new Vector();
    for (;;) {
      String line = rdr.readLine();
      if (line == null) break;
      s.addElement(line);
    }
    String[] a = new String[s.size()];
    s.copyInto(a);
    return a;
  }

  public static void main(String[] argv) throws IOException {
    String filea = argv[argv.length - 2];
    String fileb = argv[argv.length - 1];
    String[] a = slurp(filea);
    String[] b = slurp(fileb);
    Diff d = new Diff(a,b);
    char style = 'n';
    d.heuristic = false;
    for (int i = 0; i < argv.length - 2; ++i) {
      String f = argv[i];
      if (f.startsWith("-")) {
        for (int j = 1; j < f.length(); ++j) {
	  switch (f.charAt(j)) {
	  case 'H':	// heuristic on
	    d.heuristic = true; break;
	  case 'e':	// Ed style
	    style = 'e'; break;
	  case 'c':	// Context diff
	    style = 'c'; break;
	  case 'u':
	    style = 'u'; break;
	  }
	}
      }
    }
    boolean reverse = style == 'e';
    Diff.change script = d.diff_2(reverse);
    if (script == null)
      System.err.println("No differences");
    else {
      Base p;
      switch (style) {
      case 'e':
        p = new EdPrint(a,b); break;
      case 'c':
        p = new ContextPrint(a,b); break;
      case 'u':
        p = new UnifiedPrint(a,b); break;
      default:
        p = new NormalPrint(a,b);
      }
      p.print_header(filea,fileb);
      p.print_script(script);
    }
  }

}
